dashboard/test/controllers/courses_controller_test.rb

```diff
@@ -74,6 +74,58 @@ class CoursesControllerTest < ActionController::TestCase
     assert_redirected_to '/courses/csd-2018'
   end
 
+  test "show: redirect to latest stable version in course family for logged out user" do
+    sign_out @teacher
+    create :course, name: 'csp-2017', family_name: 'csp', version_year: '2017'
+    create :course, name: 'csp-2018', family_name: 'csp', version_year: '2018'
+
+    get :show, params: {course_name: 'csp-2017'}
+
+    assert_redirected_to '/courses/csp-2018/?redirect_warning=true'
+  end
+
+  test "show: redirect to latest stable version in course family for student" do
+    create :course, name: 'csp-2017', family_name: 'csp', version_year: '2017'
+    create :course, name: 'csp-2018', family_name: 'csp', version_year: '2018'
+
+    sign_in create(:student)
+    get :show, params: {course_name: 'csp-2017'}
+
+    assert_redirected_to '/courses/csp-2018/?redirect_warning=true'
+  end
+
+  test "show: do not redirect student to latest stable version in course family if they have progress" do
+    create :course, name: 'csp-2017', family_name: 'csp', version_year: '2017'
+    create :course, name: 'csp-2018', family_name: 'csp', version_year: '2018'
+
+    Course.any_instance.stubs(:has_progress?).returns(true)
+    sign_in create(:student)
+    get :show, params: {course_name: 'csp-2017'}
+
+    assert_response :ok
+  end
+
+  test "show: do not redirect student to latest stable version in course family if they are assigned" do
+    student = create :student
+    csp2017 = create :course, name: 'csp-2017', family_name: 'csp', version_year: '2017'
+    create :follower, section: create(:section, course: csp2017), student_user: student
+    create :course, name: 'csp-2018', family_name: 'csp', version_year: '2018'
+
+    sign_in student
+    get :show, params: {course_name: 'csp-2017'}
+
+    assert_response :ok
+  end
+
+  test "show: do not redirect teacher to latest stable version in course family" do
+    create :course, name: 'csp-2017', family_name: 'csp', version_year: '2017'
+    create :course, name: 'csp-2018', family_name: 'csp', version_year: '2018'
+
+    get :show, params: {course_name: 'csp-2017'}
+
+    assert_response :ok
+  end
+
```

Comment:
Nice test coverage!

Maybe I missed this, but if a student is assigned a unit in a course, but not the whole course does the logic still apply?  For example, if I'm assigned CSD - unit 1 version 17-18 and then try to go to the CSD 17-18 overview page, should I be redirected or stay?

Reply:
great question. if you are assigned to any unit in the course, you should be allowed to stay on that course